clash-lib/src/Clash/Normalize.hs

```diff
@@ -108,53 +114,85 @@ runNormalization
-                  0
-                  mempty       -- transformCounters Map
-                  globals
-                  supply
-                  (error $ $(curLoc) ++ "Report as bug: no curFun",noSrcSpan)
-                  0
-                  (IntMap.empty, 0)
-                  emptyVarEnv
-                  normState
-
-    normState = NormalizeState
-                  emptyVarEnv
-                  Map.empty
-                  emptyVarEnv
-                  emptyVarEnv
-                  Map.empty
-                  rcsMap
-
-normalize
-  :: [Id]
-  -> NormalizeSession BindingMap
-normalize []  = return emptyVarEnv
-normalize top = do
-  (new,topNormalized) <- unzip <$> mapM normalize' top
-  newNormalized <- normalize (concat new)
-  return (unionVarEnv (mkVarEnv topNormalized) newNormalized)
-
-normalize' :: Id -> NormalizeSession ([Id], (Id, Binding Term))
-normalize' nm = do
-  exprM <- lookupVarEnv nm <$> Lens.use bindings
+runNormalization env supply globals typeTrans peEval eval rcsMap lock entities session = do
+  normState <- NormalizeState
+    <$> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar Map.empty
+    <*> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar emptyVarEnv
+    <*> MVar.newMVar Map.empty
+    <*> MVar.newMVar rcsMap
+
+  rwState <- RewriteState
+    <$> MVar.newMVar mempty
+    <*> MVar.newMVar globals
+    <*> pure supply
+    <*> MVar.newMVar HashMap.empty
+    <*> MVar.newMVar 0
+    <*> MVar.newMVar (mempty, 0)
+    <*> MVar.newMVar emptyVarEnv
+    <*> pure lock
+    <*> pure normState
+
+  runRewriteSession rwEnv rwState session
+ where
+  rwEnv = RewriteEnv
+    { _clashEnv = env
+    , _typeTranslator = typeTrans
+    , _peEvaluator = peEval
+    , _evaluator = eval
+    , _topEntities = mkVarSet entities
+    }
+
+supplies :: Int -> Supply -> [Supply]
+supplies 0 _ = []
+supplies n s = let (s0', s1') = splitSupply s in s0' : supplies (n-1) s1'
+
+normalize :: [Id] -> NormalizeSession BindingMap
+normalize tops = do
+  q <- Monad.liftIO MS.newQ
+  traverse_ (Monad.liftIO . MS.pushL q) tops
+  binds <- MVar.newMVar (emptyVarSet, [])
+  uniq0 <- Lens.use uniqSupply
+  let ss = supplies (length tops) uniq0
+  -- one thread per top-level binding
+  Async.mapConcurrently_ (normalizeStep q binds) ss
+  mkVarEnv . snd <$> MVar.readMVar binds
+
+normalizeStep
+    :: MS.LinkedQueue Id
+    -> MVar (VarSet, [(Id, Binding Term)])
+    -> Supply
+    -> NormalizeSession ()
+normalizeStep q binds s = do
+  uniqSupply Lens..= s
+  res <- Monad.liftIO $ MS.tryPopR q
+  case res of
+      Just id' -> do
+        (bound, pairs) <- MVar.takeMVar binds
+        if not (id' `elemVarSet` bound)
+          then do
+            -- mark that we are attempting to normalize id'
+            MVar.putMVar binds (bound `extendVarSet` id', pairs)
+            pair <- normalize' id' q
+            MVar.modifyMVar_ binds (pure . second (pair:))
+          else
+            MVar.putMVar binds (bound, pairs)
```

Comment:
This still has the race condition. Modifying `binds` should be an atomic operation.

Reply:
I don't follow - the `takeMVar` ... `putMVar` combination holds the `MVar`, no?

We do want the `MVar.modifyMVar_ binds (pure . second (pair:))` to be separate.

Reply:
Ah sorry, you're right! I confused myself.